stablehlo/transforms/StablehloRefineParameters.cpp

```diff
@@ -0,0 +1,210 @@
+          builder.getNamedAttr("indices_of_shape_operands",
+                               builder.getI64TensorAttr({1}))});
+}
+
+ParseResult parseRefinedTypes(ModuleOp module,
+                              ArrayRef<std::string> shapeString,
+                              SmallVector<Type>& refinedTypes) {
+  MLIRContext* context = module.getContext();
+  for (const auto& shape : shapeString) {
+    Type type = parseType(shape, context);
+    if (!type) return module->emitOpError("Invalid type string: ") << shape;
+    refinedTypes.push_back(type);
+  }
+  return success();
+}
+
+// Validates refinement types:
+//   - A type refinement must be specified for each operand
+//   - Refinement types that match operand types are skipped
+//   - Refinement types that do not match operands must be refining tensors
+//   - Refined tensor types must be ranked, operand type can be unranked
+//   - Refined tensor types must match operand type for all static dimensions
+//
+LogicalResult validateRefinedTypes(func::FuncOp func, TypeRange refinedTypes) {
+  // Validate refined shapes
+  if (func.getNumArguments() != refinedTypes.size()) {
+    return func.emitOpError(
+               "number of refinements must match number of function operands ")
+           << refinedTypes.size() << " vs " << func.getNumArguments();
+  }
+
+  // Validate that refinements are valid
+  for (auto [type, refinedType] :
+       llvm::zip(func.getArgumentTypes(), refinedTypes)) {
+    // Always allow skipping refinement
+    if (type == refinedType) continue;
+
+    // If mismatched, much be tensor types
+    auto tensorType = type.dyn_cast<TensorType>();
+    auto refinedTensorType = refinedType.dyn_cast<TensorType>();
+    if (!tensorType || !refinedTensorType) {
+      return func.emitOpError("refinement type must be tensor ")
+             << type << " -> " << refinedType;
+    }
+
+    // Refined rank cannot be unranked if mismatch
+    if (refinedType.isa<UnrankedTensorType>()) {
+      return func.emitOpError("refinement must be ranked ")
+             << type << " -> " << refinedType;
+    }
+
+    if (tensorType.hasRank()) {
+      // Validate ranks match if ranked (must allow unranked tensorType)
+      if (tensorType.getRank() != refinedTensorType.getRank()) {
+        return func.emitOpError("refinement rank must match operand rank ")
+               << type << " -> " << refinedType;
+      }
+
+      // Validate static dimension sizes match
+      for (auto [dimSize, refinedDimSize] :
+           llvm::zip(tensorType.getShape(), refinedTensorType.getShape())) {
+        if (!ShapedType::isDynamic(dimSize) && dimSize != refinedDimSize) {
+          return func.emitOpError(
+                     "refinement dimension sizes must match for static "
+                     "dimensions ")
+                 << type << " -> " << refinedType;
+        }
+      }
+    }
+  }
+  return success();
+}
+
+void wrapRefinedOperands(func::FuncOp func, TypeRange refinedTypes) {
```

Comment:
A comment would be good to understand how the "operand barrier" custom calls helps in keeping the IR valid when we update the function input types to static.

Reply:
Ah just realized that we already explained that in the PR description. Thanks!
Maybe we could add a concise code comment along similar lines?

Reply:
Added description, good call